File: functional-programming/functionalProgramming.js
function main() {
  console.log("Language: JavaScript");

  const numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10];

  // --- Lambdas (arrow functions) ---
  const double = (x) => x * 2;
  const isEven = (x) => x % 2 === 0;
  const add = (a, b) => a + b;
  console.log("Arrow fn double(5):  ", double(5));

  // --- map ---
  const doubled = numbers.map(double);
  console.log("map (double):        ", doubled);

  // --- filter ---
  const evens = numbers.filter(isEven);
  console.log("filter (evens):      ", evens);

  // --- reduce ---
  const total = numbers.reduce(add, 0);
  console.log("reduce (sum):        ", total);

  // --- Composed pipeline (method chaining) ---
  const result = numbers
    .filter(isEven)
    .map(double)
    .reduce(add, 0);
  console.log("filter→map→reduce:  ", result); // even numbers, doubled, summed

  // --- Immutability patterns ---
  // const prevents rebinding; spread creates shallow copies
  const original = Object.freeze([1, 2, 3]);
  const extended = [...original, 4, 5]; // new array — original unchanged
  console.log("Frozen original:     ", original);
  console.log("Extended (new):      ", extended);

  // Object.freeze for shallow immutable objects
  const config = Object.freeze({ host: "localhost", port: 8080 });
  const updated = { ...config, port: 9090 }; // new object
  console.log("Frozen config:       ", config);
  console.log("Updated (new):       ", updated);

  console.log("Original numbers unchanged:", numbers);
}

main();

File: functional-programming/functional_programming.py
from functools import reduce
from typing import Callable


def main() -> None:
    print("Language: Python")

    numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

    # --- Lambdas ---
    double: Callable[[int], int] = lambda x: x * 2
    is_even: Callable[[int], bool] = lambda x: x % 2 == 0
    add: Callable[[int, int], int] = lambda a, b: a + b
    print("Lambda double(5):   ", double(5))

    # --- map ---
    doubled = list(map(double, numbers))
    print("map (double):       ", doubled)

    # --- filter ---
    evens = list(filter(is_even, numbers))
    print("filter (evens):     ", evens)

    # --- reduce ---
    total = reduce(add, numbers)
    print("reduce (sum):       ", total)

    # --- Composed pipeline ---
    result = reduce(add, map(double, filter(is_even, numbers)))
    print("filter→map→reduce: ", result)  # even numbers, doubled, summed

    # --- Immutability patterns ---
    # Tuples are immutable; list operations return new lists, originals unchanged
    original = (1, 2, 3)
    extended = original + (4, 5)           # new tuple — original unchanged
    print("Immutable tuple original:  ", original)
    print("Immutable tuple extended:  ", extended)

    # list comprehension returns new list
    squared = [x ** 2 for x in numbers]
    print("List comprehension (new):  ", squared)
    print("Original unchanged:        ", numbers)


if __name__ == "__main__":
    main()
